File: 프로그래머스/1/42840. 모의고사/모의고사.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

class Solution {
	private static final int[] PATTERN_ONE = {1, 2, 3, 4, 5};
	private static final int[] PATTERN_TWO = {2, 1, 2, 3, 2, 4, 2, 5};
	private static final int[] PATTERN_THREE = {3, 3, 1, 1, 2, 2, 4, 4, 5, 5};

	public int[] solution(int[] answers) {
		List<Integer> scores = new ArrayList<>();

		scores.add(evaluate(PATTERN_ONE, answers));
		scores.add(evaluate(PATTERN_TWO, answers));
		scores.add(evaluate(PATTERN_THREE, answers));

		int maxScore = scores.stream().max(Comparator.naturalOrder()).orElse(0);
		List<Integer> result = new ArrayList<>();

		for (int i = 1; i <= 3; i++) {
			if (scores.get(i - 1) == maxScore) {
				result.add(i);
			}
		}

		return result.stream().mapToInt(Integer::intValue).toArray();
	}

	public int evaluate(int[] pattern, int[] answers) {
		int score = 0;

		for (int i = 0; i < answers.length; i++) {
			int patternIdx = i % pattern.length;
			if (pattern[patternIdx] == answers[i]) {
				score++;
			}
		}

		return score;
	}
}

File: 프로그래머스/1/42840. 모의고사/모의고사.py
from typing import List

def solution(answers: List[int]) -> List[int]:
    PATTERN_ONE = [1,2,3,4,5]
    PATTERN_TWO = [2,1,2,3,2,4,2,5]
    PATTERN_THREE = [3,3,1,1,2,2,4,4,5,5]
    
    scores = [0, 0, 0]
    
    for i, answer in enumerate(answers):
        if PATTERN_ONE[i % len(PATTERN_ONE)] == answer:
            scores[0] += 1
        if PATTERN_TWO[i % len(PATTERN_TWO)] == answer:
            scores[1] += 1
        if PATTERN_THREE[i % len(PATTERN_THREE)] == answer:
            scores[2] += 1
    
    max_scores = max(scores)
    
    return [i+1 for i, score in enumerate(scores) if score == max_scores]